docs/getting_started/how_to_build_an_api.md

```diff
@@ -0,0 +1,215 @@
+- it makes an initial call to an external endpoint to get exchange rates, and stores them in a kvstore
+- it occasionally makes more calls to obtain exchange rates, and keep the kvstore updates
+
+The _"app"_ module is our api server, notice that it was created with a call to `apiserver.New`:
+
+[embedmd]:# (../../examples/getting_started/api/main.go /app.Add\(\"api\"/ /\)/)
+```go
+app.Add("api", apiserver.New(definer.ApiDefiner)
+```
+
+The code for _ApiDefiner_ is as follows:
+
+[embedmd]:# (../../examples/getting_started/api/definer/definer.go /func ApiDefiner/ /\n}/)
+```go
+func ApiDefiner(ctx context.Context, config cfg.Config, logger log.Logger) (*apiserver.Definitions, error) {
+	definitions := &apiserver.Definitions{}
+
+	currencyService, err := currency.New(ctx, config, logger)
+	if err != nil {
+		return nil, fmt.Errorf("can not create currencyService: %w", err)
+	}
+
+	definitions.GET("/euro/:amount/:currency", apiserver.CreateHandler(handler.NewEuroHandler(logger, currencyService)))
+	definitions.GET("/euro-at-date/:amount/:currency/:date", apiserver.CreateHandler(handler.NewEuroAtDateHandler(logger, currencyService)))
+
+	return definitions, nil
+}
+```
+
+_ApiDefiner_ does four things:
+
+1. Creates a _currencyService_ which offers currency exchange functionality by using the kvstore.
+2. It creates a new EuroHandler.
+3. Creates a new EuroAtHandler.
+4. It instantiates a new _api.Definitions_ objects to which it adds to routes served by the two handlers.
+
+Both EuroHandler and EuroAtDateHandler to similar things: the first takes in a money amount and a currency, and returns its value in euro, while the other takes in a date as well, and returns the euro value using the exchange rate for that given date.
+
+[embedmd]:# (../../examples/getting_started/api/handler/handler.go /type euroHandler struct/ /\n}/)
+```go
+type euroHandler struct {
+	logger          log.Logger
+	currencyService currency.Service
+}
+```
+
+The _euroHandlerStruct_ has a _logger_ and a _currencyService_ field. It is a private struct, but there is a convenient way to instantiate one:
+
+[embedmd]:# (../../examples/getting_started/api/handler/handler.go /func NewEuroHandler/ /\)/)
+```go
+func NewEuroHandler(logger log.Logger, currencyService currency.Service)
+```
+
+The most important method of _euroHanlder_ is _Handle_:
+
+[embedmd]:# (../../examples/getting_started/api/handler/handler.go /func \(h \*euroHandler\) Handle/ /\n}/)
+```go
+func (h *euroHandler) Handle(requestContext context.Context, request *apiserver.Request) (response *apiserver.Response, error error) {
+	currency := request.Params.ByName("currency")
+	amountString := request.Params.ByName("amount")
+	amount, err := strconv.ParseFloat(amountString, 64)
+	if err != nil {
+		h.logger.Error("cannot parse amount %s: %w", amountString, err)
+
+		return apiserver.NewStatusResponse(http.StatusInternalServerError), err
```

Comment:
If you return an error, the response you return is ignored, so this is already a bad example in the docs. Also, 500 Internal Server Error is not the correct response if the client sent a bad request, that would be 400 bad request